src/sinks/gcp/cloud_storage.rs

```diff
@@ -0,0 +1,543 @@
+    storage_class: HeaderValue,
+    tags: Vec<(HeaderName, HeaderValue)>,
+    extension: String,
+    time_format: String,
+    append_uuid: bool,
+}
+
+impl RequestSettings {
+    fn new(config: &GcsSinkConfig) -> crate::Result<Self> {
+        let acl = HeaderValue::from_str(&to_string(&config.acl)).unwrap();
+        let content_type = HeaderValue::from_str(config.encoding.content_type()).unwrap();
+        let content_encoding = config
+            .compression
+            .content_encoding()
+            .map(|ce| HeaderValue::from_str(&to_string(ce)).unwrap());
+        let storage_class = HeaderValue::from_str(&to_string(&config.storage_class)).unwrap();
+        let tags = config
+            .tags
+            .as_ref()
+            .map(|tags| tags.iter().map(make_header).collect::<Result<Vec<_>, _>>())
+            .unwrap_or(Ok(vec![]))?;
+        let extension = config
+            .filename_extension
+            .clone()
+            .unwrap_or_else(|| config.compression.extension().into());
+        let time_format = config.filename_time_format.clone().unwrap_or("%s".into());
+        let append_uuid = config.filename_append_uuid.unwrap_or(true);
+        Ok(Self {
+            acl,
+            content_type,
+            content_encoding,
+            storage_class,
+            tags,
+            extension,
+            time_format,
+            append_uuid,
+        })
+    }
+}
+
+// Make a header pair from a key-value string pair
+fn make_header(kv: (&String, &String)) -> crate::Result<(HeaderName, HeaderValue)> {
+    let (name, value) = kv;
+    Ok((
+        HeaderName::from_bytes(name.as_bytes())?,
+        HeaderValue::from_str(&value)?,
+    ))
+}
+
+fn encode_event(
+    event: Event,
+    key_prefix: &Template,
+    encoding: &Encoding,
+) -> Option<PartitionInnerBuffer<Vec<u8>, Bytes>> {
+    let key = key_prefix
+        .render_string(&event)
+        .map_err(|missing_keys| {
+            warn!(
+                message = "Keys do not exist on the event. Dropping event.",
+                ?missing_keys,
+                rate_limit_secs = 30,
+            );
+        })
+        .ok()?;
+
+    let log = event.into_log();
+    let bytes = match encoding {
+        Encoding::Ndjson => serde_json::to_vec(&log.unflatten())
+            .map(|mut b| {
+                b.push(b'\n');
+                b
+            })
+            .expect("Failed to encode event as json, this is a bug!"),
+        &Encoding::Text => {
+            let mut bytes = log
+                .get(&event::log_schema().message_key())
+                .map(|v| v.as_bytes().to_vec())
+                .unwrap_or_default();
+            bytes.push(b'\n');
+            bytes
+        }
+    };
+
+    Some(PartitionInnerBuffer::new(bytes, key.into()))
+}
+
+#[derive(Clone)]
+pub struct GcsRetryLogic;
+
+// This is a clone of HttpRetryLogic for the Body type, should get merged
+impl RetryLogic for GcsRetryLogic {
```

Comment:
What is different here and httpretrylogic? This again leads me to think we should implement this in terms of `HttpSink`?

Reply:
`HttpRetryLogic` has `type Response = hyper::Response<Bytes>` while this one uses `<Body>`. I was considering parameterizing `HttpRetryLogic` on that inner type, but didn't get around to it.

Reply:
body is basically a wrapper for bytes so may make sense to find a nice middle ground here. We _also_ could create our own body type to make things easier.